### PR TITLE
[VPN 2.0] Refresh VPN subscriber credential on expiry

### LINKED ISSUE #56692: Implement purchased state management for BraveVpnService v2
We need to implement purchased state management for BraveVpnService v2, similar to what we have in v1.
This is mostly shared between desktop and Android. The agreement is that we still keep subscriber credential in the browser service, and pass it to the agent/helper whenever appropriate.
Acceptance criteria:
- Purchased state is properly fetched and maintained
- SKUS credential is updated whenever appropriate
- Subscriber credential is updated whenever appropriate
- Behaviour from the customer's perspective is similar to v1.

### PR DESCRIPTION
**Pull request by alekspop26**
Schedule a one-shot timer whenever a subscriber credential is cached:
- at startup;
- on a successful exchange;
- on the cached-credential fast path.

The refresh call clears the credential and re-runs the reload chain when the subscriber credential expires, so an active subscription keeps a live credential without user action. Unsuccessful refresh doesn't retry automatically and relies on a regular UI action initiated by a user.

The change also simplifies CheckInitialState logic, thus removing one of the potential `ScheduleSubscriberCredentialRefresh` call sites.

Resolves https://github.com/brave/brave-browser/issues/56692

### CODE REVIEW

File: components/brave_vpn/browser/v2/purchased_state_manager.cc
Comment (szilardszaloki) [inline components/brave_vpn/browser/v2/purchased_state_manager.cc:1]: Once everything goes through the endpoint client, you could use `request.timeout_duration` instead of an explicit timer. That's per-request rather than per-load, though, unlike `load_timeout_timer_`.

File: components/brave_vpn/browser/v2/purchased_state_manager.cc [hunk 1]
```diff
@@ -456,8 +458,29 @@ void PurchasedStateManager::RunPurchasedStateCallback(
   purchased_state_changed_callback_.Run(state, std::move(description));
 }
 
-void PurchasedStateManager::ScheduleSubscriberCredentialRefresh() {
-  NOTIMPLEMENTED();
+void PurchasedStateManager::ScheduleSubscriberCredentialRefresh(
+    const base::Time& expiration_time) {
+  if (subscriber_credential_refresh_timer_.IsRunning()) {
+    subscriber_credential_refresh_timer_.Stop();
+  }
```
Comment (szilardszaloki) [inline components/brave_vpn/browser/v2/purchased_state_manager.cc]: I think that's redundant, calling `Start()` directly on a running timer should be supported and do the same thing.

File: components/brave_vpn/browser/v2/credential_store.cc [hunk 2]
```diff
@@ -46,10 +47,21 @@ bool IsCredentialValid(const base::DictValue& dict,
...
 }
 
-std::string GetCredential(const base::DictValue& dict,
-                          std::string_view credential_key) {
-  const std::string* credential = dict.FindString(credential_key);
-  return credential ? *credential : std::string();
+// Reads the keyed credential and the shared expiration from the slot as one
+// bundle, or nullopt if the credential is not valid. IsCredentialValid()
+// verifies - on this same dict - that the credential string and a parseable,
+// future expiration are both present, so the dereferences below cannot fail.
```
Comment (netzenbot) [inline components/brave_vpn/browser/v2/credential_store.cc]: nit: "below" is a positional reference that goes stale if anything is inserted between the comment and the code. Name the expressions instead, e.g. "so the `FindString()` / `ValueToTime()` dereferences in the returned `Credential` cannot fail." ([❌ Don't Use Positional Terms in Code Comments](https://github.com/brave/brave-core/tree/master/docs/best-practices/coding-standards.md#CS-024))

File: components/brave_vpn/browser/v2/purchased_state_manager_unittest.cc [hunk 3]
```diff
@@ -1061,6 +1073,47 @@ TEST_F(PurchasedStateManagerTest, ExpiredSummaryMeansNotPurchased) {
...
+// re-resolves from scratch (a full summary round-trip, visible as LOADING).
+TEST_F(PurchasedStateManagerTest, SuccessSchedulesRefreshThatReresolves) {
+  CreateManager();
+  const base::Time expiry = base::Time::Now() + base::Days(30);
+  SeedSkusCredential(expiry);
+  manager_->Load(CurrentDomain());
+
+  api_client_.ResolveExchange(base::ok(std::string(kTestSubscriberCredential)));
+  ASSERT_TRUE(manager_->IsPurchased());
+  EXPECT_TRUE(RefreshScheduled());
```
Comment (netzenbot) [inline components/brave_vpn/browser/v2/purchased_state_manager_unittest.cc:1098]: `ScheduleSubscriberCredentialRefresh()` now has three call sites, but only the post-exchange one is asserted here. Since `RefreshScheduled()`/`RefreshDelay()` already exist, adding `EXPECT_TRUE(RefreshScheduled())` to `InitialStateWithValidCredentialIsPurchased` and `CachedCredentialFastPathCancelsSilentLoad` would cover the restart-from-prefs path, which is the common real-world case. ([✅ PRs Should Include Reasonable Test Coverage](https://github.com/brave/brave-core/tree/master/docs/best-practices/testing-isolation.md#TI-039))

Comment (alekspop26) [reply components/brave_vpn/browser/v2/purchased_state_manager_unittest.cc:1098]: Fixed.

File: components/brave_vpn/browser/v2/credential_store.cc (hunk 2 above)
Comment (alekspop26) [reply components/brave_vpn/browser/v2/credential_store.cc]: Not related to this CR.

File: components/brave_vpn/browser/v2/purchased_state_manager.cc (hunk 1 above)
Comment (alekspop26) [reply components/brave_vpn/browser/v2/purchased_state_manager.cc]: Fixed.

File: components/brave_vpn/browser/v2/purchased_state_manager.cc [hunk 4]
```diff
@@ -456,8 +458,25 @@ void PurchasedStateManager::RunPurchasedStateCallback(
...
+                     base::Unretained(this)));
+}
+
+void PurchasedStateManager::RefreshSubscriberCredential() {
+  VLOG(2) << "Refreshing subscriber credential...";
+
+  // Clear the cached credential to get a new subscriber credential.
+  credential_store_->Clear();
+  Reload();
 }
```
Comment (szilardszaloki) [inline components/brave_vpn/browser/v2/purchased_state_manager.cc:485]: Should `RefreshSubscriberCredential()` guard against a load already being in flight before it `Clear()`s + `Reload()`s?

File: components/brave_vpn/browser/v2/purchased_state_manager.cc [hunk 5]
```diff
@@ -159,11 +159,13 @@ void PurchasedStateManager::SetPurchasedState(
...
 
 void PurchasedStateManager::CheckInitialState() {
```
Comment (szilardszaloki) [inline components/brave_vpn/browser/v2/purchased_state_manager.cc:165]: Two of `CheckInitialState()`'s three branches look like they duplicate logic that `Load()` already has:

- valid subscriber credential `==>` `PURCHASED` + schedule refresh (same as `Load()`)
- valid SKUS credential `==>` reload/exchange (same as `Load()`)

So `CheckInitialState()` could potentially shrink to just delegating to `Reload()`, keeping only what's genuinely startup-specific?
```cpp
void PurchasedStateManager::CheckInitialState() {
  if (!credential_store_->HasAnyCredential()) {
    return;
  }

  if (!credential_store_->GetValidSubscriberCredential() &&
      !credential_store_->GetValidSkusCredential()) {
    credential_store_->Clear();
  }

  Reload();
}
```

Comment (szilardszaloki) [reply components/brave_vpn/browser/v2/purchased_state_manager.cc:165]: Actually, went ahead and verified locally — all 40 `PurchasedStateManagerTest.*` pass with this.

Comment (szilardszaloki) [reply components/brave_vpn/browser/v2/purchased_state_manager.cc:165]: None of this needs to happen in this PR. But the file does seem to carry a bit of incidental complexity here and there, so it'd be nice to simplify and trim it where we can, either now or in a follow-up.

**Review by szilardszaloki — APPROVED** [review]
Commit `[VPN 2.0] Refresh VPN subscriber credential on expiry` LGTM — modulo comments.

File: components/brave_vpn/browser/v2/purchased_state_manager.cc (hunk 4 above)
Comment (alekspop26) [reply components/brave_vpn/browser/v2/purchased_state_manager.cc:485]: Good callout, thanks! Added protection for in-flight loads, both visible and silent ones.

File: components/brave_vpn/browser/v2/purchased_state_manager.cc (hunk 5 above)
Comment (alekspop26) [reply components/brave_vpn/browser/v2/purchased_state_manager.cc:165]: This is a great simplification, thanks for pointing that out! It is sort of related to this PR, as it removes one of the (unneeded) call sites for `ScheduleSubscriberCredentialRefresh`, so I decided to include the change. Also, added one more unit test to cover an edge case related to a present SKUS credential at startup.

File: components/brave_vpn/browser/v2/purchased_state_manager.cc
Comment (alekspop26) [reply components/brave_vpn/browser/v2/purchased_state_manager.cc:1]: Thanks, but I'd keep the explicit timer here, exactly for the reason you mentioned: it's a per-load timeout cancellation.

File: components/brave_vpn/browser/v2/purchased_state_manager.h [hunk 6]
```diff
@@ -121,6 +123,8 @@ class PurchasedStateManager final {
   std::string loading_environment_;
   uint64_t loading_sequence_ = 0;
   base::OneShotTimer load_timeout_timer_;
+  base::OneShotTimer subscriber_credential_refresh_timer_;
+  bool subscriber_credential_refresh_deferred_ = false;
```
Comment (szilardszaloki) [inline components/brave_vpn/browser/v2/purchased_state_manager.h]: That `bool` feels like a fair bit of coordination (set in one place, cleared in two, consumed+cleared in a third), and IIUC it only actually comes into play when **all of the following are true**: you're `PURCHASED` with a valid credential, that credential's refresh becomes due, an other-env load is in flight at that exact moment, **and** that load fails. That can't happen in prod, and even elsewhere the failure it avoids seems fairly soft — you end up with a stale `PURCHASED` that fixes itself on the next user interaction? So I mostly just want to double-check that the added complexity is worth it.

If we keep it, `std::optional<std::string> deferred_refresh_env_;` (the env the refresh is owed for, checked against the current env when consumed) would enforce the "don't replay across a switch" invariant at the point of use (and also gets rid of having to remember the `= false` in `OnPrepareCredentialsPresentation()`).

If we ever wanted to take it a step further, making refresh and other-env load mutually exclusive states (e.g. some `enum class { kSettled, kMigrating };`, stop the timer on entering `kMigrating`, re-arm it on exit) removes the coordination entirely, as refresh simply can't fire during a migration because those states can't coexist. Bigger change though, not suggesting that for this PR.

Comment (alekspop26) [reply components/brave_vpn/browser/v2/purchased_state_manager.h]: Fair point, the deferral is not worth the added complexity I guess. Let's just keep the guard.

Comment (alekspop26) [reply components/brave_vpn/browser/v2/purchased_state_manager.h]: Reverted the deferral and its new tests.
